cpython/patches/0005-Use-wasm-gc-based-call-adaptor-if-available.patch

```diff
@@ -0,0 +1,381 @@
++
++// Enable macro expanding in the body of EM_JS
++#define EM_JS_MACROS(ret, func_name, args, body...)                            \
++  EM_JS(ret, func_name, args, body)
++
++EM_JS_MACROS(CountArgsFunc, _PyEM_GetCountArgsPtr, (), {
++  return Module._PyEM_CountArgsPtr; // initialized below
++}
++// Binary module for the checks. It has to be done in web assembly because
++// clang/llvm have no support yet for the reference types yet. In fact, the wasm
++// binary toolkit doesn't yet support the ref.test instruction either. To
++// convert the following module to the binary, my approach is to find and
++// replace "ref.test $type" -> "drop i32.const n" on the source text. This
++// results in the bytes "0x1a, 0x41, n" where we need the bytes "0xfb, 0x14, n"
++// so doing a find and replace on the output from "0x1a, 0x41" -> "0xfb, 0x14"
++// gets us the output we need.
++//
++// (module
++//     (type $type0 (func (param) (result i32)))
++//     (type $type1 (func (param i32) (result i32)))
++//     (type $type2 (func (param i32 i32) (result i32)))
++//     (type $type3 (func (param i32 i32 i32) (result i32)))
++//     (type $type4 (func (param i32 i32 i32 i32) (result i32)))
++//     (type $blocktype (func (param i32) (result)))
++//     (table $funcs (import "e" "t") 0 funcref)
++//     (export "f" (func $f))
++//     (func $f (param $fptr i32) (result i32)
++//              (local $fref funcref)
++//         local.get $fptr
++//         table.get $funcs
++//         local.tee $fref
++//         ref.test $type4
++//         (block $b (type $blocktype)
++//             i32.eqz
++//             br_if $b
++//             i32.const 4
++//             return
++//         )
++//         local.get $fref
++//         ref.test $type3
++//         (block $b (type $blocktype)
++//             i32.eqz
++//             br_if $b
++//             i32.const 3
++//             return
++//         )
++//         local.get $fref
++//         ref.test $type2
++//         (block $b (type $blocktype)
++//             i32.eqz
++//             br_if $b
++//             i32.const 2
++//             return
++//         )
++//         ref.test $type1
++//         i32.const 1
++//         (block $b (type $blocktype)
++//             i32.eqz
++//             br_if $b
++//             i32.const 1
++//             return
++//         )
++//         ref.test $type0
++//         i32.const 0
++//         (block $b (type $blocktype)
++//             i32.eqz
++//             br_if $b
++//             i32.const 0
++//             return
++//         )
++//         i32.const -1
++//     )
++// )
++addOnPreRun(() => {
++  // Try to initialize countArgsFunc
++  const code = new Uint8Array([
++    0x00, 0x61, 0x73, 0x6d, // \0asm magic number
++    0x01, 0x00, 0x00, 0x00, // version 1
++    0x01, 0x23, // Type section, body is 0x23 bytes
++      0x06, // 6 entries
++      0x60, 0x00, 0x01, 0x7f,                         // (type $type0 (func (param) (result i32)))
```

Comment:
This part is quite scary 😅

Reply:
Well you can see the wat above. The problem is that the wasm binary toolkit doesn't know how to handle `ref.test $type4`. Probably I should make a wasm binary toolkit PR to add support for this, but we'll see if I get around to it.